### PR TITLE
Chart lib cleanup

### PR DESCRIPTION
**Pull request by camsaul**
Attempt to make LineAreaBarScatterRenderer something that's actually possbile to work on.

*  Namespace reduced from 1500+ LOC to 644
*  `lineAreaBar()` reduced from 500+ LOC to ~70

This was accomplished by breaking tons of things out into separate functions and and breaking the namespace up into a few different parts. The underlying logic is unchanged.

Still planning on adding more Flow type checking stuff and documentation but take a look and let me know if this makes sense.

### CODE REVIEW

**Review by kdoh — APPROVED** [review]
Definitely feels easier to parse.

File: frontend/src/metabase/visualizations/lib/RowRenderer.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    const dimension = dataset.dimension(d => d[0]);
+    const group = dimension.group().reduceSum(d => d[1]);
+    const xDomain = d3.extent(rows, d => d[1]);
+    const yValues = rows.map(d => d[0]);
+
+    forceSortedGroup(group, makeIndexMap(yValues));
+
+    initChart(chart, element);
+
+    chart.on("renderlet.tooltips", chart => {
```
Comment (salsakran) [inline frontend/src/metabase/visualizations/lib/RowRenderer.js]: should this be woven into the tooltips file?

File: frontend/src/metabase/visualizations/lib/LineAreaBarPostRenderer.js [hunk 2]
```diff
@@ -0,0 +1,323 @@
+import d3 from "d3";
+
```
Comment (salsakran) [inline frontend/src/metabase/visualizations/lib/LineAreaBarPostRenderer.js]: I'm not sure "PostRenderer" makes sense as these are a collection of functions that should be run post-rendering. Maybe "PostRendering"?

File: frontend/src/metabase/visualizations/lib/LineAreaBarPostRenderer.js [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    beforeRenderHideDisabledAxesAndLabels(chart, settings);
+    beforeRenderFixMargins(chart, settings);
+}
+
+// +-------------------------------------------------------------------------------------------------------------------+
+// |                                              PUTTING IT ALL TOGETHER                                              |
+// +-------------------------------------------------------------------------------------------------------------------+
+
+/// once chart has rendered and we can access the SVG, do customizations to axis labels / etc that you can't do through dc.js
+export default function lineAndBarOnRender(chart, settings, onGoalHover, isSplitAxis, isStacked) {
```
Comment (salsakran) [inline frontend/src/metabase/visualizations/lib/LineAreaBarPostRenderer.js]: This is a weird function name ... what exactly is it supposed to be doing?

File: frontend/src/metabase/visualizations/lib/LineAreaBarRenderer.js [hunk 4]
```diff
@@ -8,1013 +8,176 @@ import _ from "underscore";
...
+
+    return {
+        xValues,
+        xDomian: d3.extent(xValues),
+        xInterval: getXInterval(props, xValues)
+    };
+}
+
+
+function fillMissingValuesInDatas(props, { xValues, xDomain, xInterval }, datas) {
```
Comment (salsakran) [inline frontend/src/metabase/visualizations/lib/LineAreaBarRenderer.js]: this would be a good candidate to pull out of this file as well

**Review by salsakran — COMMENTED** [review]
Looks like a reasonable first split. I think it would be useful to try to pull out things like tooltips, axis control, missing value interpolation etc that is independent of the chart type into its own function namespaces, and separate data manipulation from d3 futzing from dom manipulation.

File: frontend/src/metabase/visualizations/lib/LineAreaBarPostRenderer.js (hunk 3 above)
Comment (camsaul) [reply frontend/src/metabase/visualizations/lib/LineAreaBarPostRenderer.js]: I didn't change the name of this function, just moved it to a different file. My understanding is that it is the `onRender` function that gets called for line *and* bar charts (as well as area and scatter charts)

**Pull request merged**